ufl/algorithms/renumbering.py

```diff
@@ -1,87 +1,71 @@
-    please do not use for anything else.
-    """
+    expr = MultiFunction.reuse_if_untouched
 
-    def __init__(self):
-        """Initialise."""
-        VariableRenumberingTransformer.__init__(self)
-        self.index_map = {}
+    def multi_index(self, o):
+        """Apply to multi-indices."""
+        return type(o)(
+            tuple(self.index_cache[i] if isinstance(i, Index) else i for i in o.indices())
+        )
 
     def zero(self, o):
         """Apply to zero."""
         fi = o.ufl_free_indices
         fid = o.ufl_index_dimensions
-        mapped_fi = tuple(self.index(Index(count=i)) for i in fi)
-        paired_fid = [(mapped_fi[pos], fid[pos]) for pos, a in enumerate(fi)]
-        new_fi, new_fid = zip(*tuple(sorted(paired_fid)))
-        return Zero(o.ufl_shape, new_fi, new_fid)
-
-    def index(self, o):
-        """Apply to index."""
-        if isinstance(o, FixedIndex):
+        new_indices = [self.index_cache[Index(i)].count() for i in fi]
+        if fi == () and fid == ():
             return o
-        else:
-            c = o._count
-            i = self.index_map.get(c)
-            if i is None:
-                i = Index(count=len(self.index_map))
-                self.index_map[c] = i
-            return i
+        new_fi, new_fid = zip(*sorted(zip(new_indices, fid), key=lambda x: x[0]))
+        return type(o)(o.ufl_shape, tuple(new_fi), tuple(new_fid))
 
-    def multi_index(self, o):
-        """Apply to multi_index."""
-        new_indices = tuple(self.index(i) for i in o.indices())
-        return MultiIndex(new_indices)
 
+def renumber_indices(form):
+    """Renumber indices to have a consistent index numbering starting from 0.
 
-def renumber_indices(expr):
-    """Renumber indices."""
-    if isinstance(expr, Expr):
-        num_free_indices = len(expr.ufl_free_indices)
+    This is useful to avoid multiple kernels for the same integrand,
+    but with different subdomain ids.
 
-    result = apply_transformer(expr, IndexRenumberingTransformer())
+    Args:
+        form: A UFL form, integral or expression.
 
-    if isinstance(expr, Expr):
-        if num_free_indices != len(result.ufl_free_indices):
-            raise ValueError(
-                "The number of free indices left in expression "
-                "should be invariant w.r.t. renumbering."
-            )
-    return result
+    Returns:
+        A new form, integral or expression with renumbered indices.
+    """
+    if isinstance(form, Form):
+        new_integrals = [renumber_indices(itg) for itg in form.integrals()]
+        return Form(new_integrals)
+    elif isinstance(form, Integral):
+        integral = form
+        reindexer = IndexRelabeller()
+        new_integrand = map_expr_dag(reindexer, integral.integrand())
+        return integral.reconstruct(new_integrand)
+    elif isinstance(form, Expr):
+        expr = form
+        reindexer = IndexRelabeller()
+        return map_expr_dag(reindexer, expr)
+    else:
```

Comment:
Is this not exactly what `map_integrand_dags` is for? I forget.

I'll note that the index renumbering pass here is still a "bad" algorithm, though it is at least dag aware. In the following way: index numbers only matter for free indices, not bound indices, because `A = ComponentTensor(A[i, j], (i, j))` and `B = ComponentTensor(A[k, l], (k, l))` are identical expressions. It might be that what you want this for, this is not a problem. There is some discussion of this in section 9.2.5 of Miklos' thesis: https://spiral.imperial.ac.uk/bitstream/10044/1/62636/1/Homolya-M-2018-PhD-Thesis.pdf

Reply:
I am not sure I follow your first comment here.
I am following the same logic as in:
https://github.com/FEniCS/ufl/blob/60a6956a553a014d9b9dc5fbd66c7e764dea0f72/ufl/algorithms/apply_geometry_lowering.py#L473-L497

For the second comment: For the issue at hand, I've only encountered issues with free indices, so I am not concerned with bound indices for now.

Reply:
> I am not sure I follow your first comment here.

https://github.com/FEniCS/ufl/blob/main/ufl/algorithms/map_integrands.py

`map_integrands` (called via `map_integrand_dags`) does (in a more systematic way) this handling of different types of input.

Reply:
It doesn't really matter I suppose.

Reply:
It does look a whole lot neater with `map_integrand_dags`!